sdk/typescript/telemetry/telemetry.ts

```diff
@@ -0,0 +1,222 @@
+ */
+function otelEnabled(name?: string): boolean {
+  if (name) {
+    name = `OTEL_TYPESCRIPT_${name.toUpperCase()}_INSTRUMENTATION_DISABLED`
+  }
+
+  if (!name) {
+    name = OtelEnv.OTEL_SDK_DISABLED
+  }
+
+  return process.env[name] !== "true"
+}
+
+/**
+ * A wrapper around the OpenTelemetry SDK to configure it for Dagger.
+ */
+class DaggerOtelConfigurator {
+  private is_configured = false
+
+  private is_enabled = otelEnabled()
+
+  private sdk?: NodeSDK
+
+  /**
+   * Initialize the Open Telemetry SDK if enabled or not already configured.
+   */
+  initialize() {
+    if (this.is_configured || !this.is_enabled) {
+      return
+    }
+
+    console.debug("Initialiazing telemetry")
+    this.configure()
+
+    // Create the node SDK with the context manager, the resource and the exporter.
+    const sdk = new NodeSDK({
+      resource: new Resource({
+        [SEMRESATTRS_SERVICE_NAME]: "dagger-typescript-sdk",
+      }),
+    })
+
+    // Register the SDK to the OpenTelemetry API if OTEL is enabled
+    sdk.start()
+    console.debug("Telemetry initialized")
+    this.is_configured = true
+    this.sdk = sdk
+  }
+
+  /**
+   * Shutdown the Open Telemetry SDK to flush traces and metrics and close the connection.
+   */
+  async close() {
+    if (this.sdk) {
+      this.sdk.shutdown()
+    }
+  }
+
+  configure() {
+    if (!otelConfigured()) {
+      console.debug("Telemetry not configured")
+      return
+    }
+
+    if (!otelEnabled()) {
+      console.debug("Telemetry disabled")
+      return
+    }
+
+    this.setupEnv()
+  }
+
+  /**
+   * Setup environment for auto-configuring the SDK.
+   */
+  setupEnv() {
+    // Setup a default value for the service name.
+    if (!process.env[OtelEnv.OTEL_SERVICE_NAME]) {
+      process.env[OtelEnv.OTEL_SERVICE_NAME] = SERVICE_NAME
+    }
+
+    // TODO: The insecure flag should be set by the shim instead of the SDK.
+    const endpoints = [
+      OtelEnv.OTEL_EXPORTER_OTLP_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT,
+    ]
+
+    endpoints.forEach((endpoint) => {
+      const env = process.env[endpoint]
+      if (env && env.startsWith("http://")) {
+        process.env[OtelEnv.OTEL_EXPORTER_OTLP_INSECURE] = "true"
+      }
+    })
+  }
+}
+
+const configurator = new DaggerOtelConfigurator()
+
+export function initiliaze() {
+  configurator.initialize()
+}
+
+export async function close() {
+  await configurator.close()
+}
+
+/**
+ * Return a tracer to use with Dagger.
+ *
+ * The tracer is automatically initialized if not already done.
+ * As a conveniance function, you can use `withTracingSpan` that automatically close
+ * the span at the end of the function.
+ */
+export function getTracer(): opentelemetry.Tracer {
+  initiliaze()
+
+  const tracer = opentelemetry.trace
+    .getTracerProvider()
+    .getTracer("dagger-typescript-sdk")
```

Comment:
This is the service name. For the tracer, the convention is:
```suggestion
    .getTracer("dagger.io/sdk.typescript")
```

Reply:
done